central/sensorupgradeconfig/datastore/datastore_impl.go

```diff
@@ -57,6 +57,16 @@ func (d *dataStore) UpsertSensorUpgradeConfig(ctx context.Context, sensorUpgrade
 		return errors.New("auto-upgrade not allowed on managed central")
```

Comment:
What concerns me the most is that `env.ManagedCentral.BooleanSetting()` is used in multiple files to figure out eligibility for auto-upgrades. This means we have a leaky abstraction. A slightly better version of it could be to define a function that derives `SensorAutoUpgradeFeatureStatus` based on `env.ManagedCentral.BooleanSetting()` and use it everywhere.

However, that doesn't address my other commends, and I think, the root of the problem is that you're trying to persist `auto_upgrade_allowed` attribute in a database while in fact it is automatically derived. There's no need to persist something that's automatically derived.

To make things better you could:
1. Remove `SensorAutoUpgradeFeatureStatus auto_upgrade_allowed = 2;` from `proto/storage/sensor_upgrade.proto`, i.e. not try to persist it.
2. Make `SensorAutoUpgradeFeatureStatus` a part of `GetSensorUpgradeConfigResponse`.
3. Remove any mentions of `env.ManagedCentral.BooleanSetting()` and `auto_upgrade_allowed`-derivatives from `central/sensorupgradeconfig/datastore/datastore_impl.go`
4. Compute `SensorAutoUpgradeFeatureStatus` in `central/sensorupgrade/service/service_impl.go`.

This should leave you with a hardly testable `addDefaultConfigIfEmpty` in `central/sensorupgradeconfig/datastore/singleton.go` that needs to derive `SensorAutoUpgradeFeatureStatus` from `env.ManagedCentral.BooleanSetting()`. Let's think how to simplify that one. I have a strong suspicion that that method is actually redundant and the calculation of defaults can be done elsewhere.

Another thing that will remain in question is the default value for `dataStore.autoTrigger`. I think it can be safely pushed up to be a part of `central/sensorupgrade/service/service.go`, i.e. the _service_ contract, not the _datasource_ contract. There, in `service_impl.go` the defaults can be applied even in the case when datastore doesn't hold any value.

Reply:
That makes sense.